scripts/nn/layers/softmax2d.dml

```diff
@@ -0,0 +1,69 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/*
+ * 2D - Softmax classifier layer used for image segmentation.
+ */
+
+forward = function(matrix[double] scores, int C, int H, int W)
+    return (matrix[double] probs) {
+  /*
+   * Computes the forward pass for a softmax classifier.  The inputs
+   * are interpreted as unnormalized, log-probabilities for each of
+   * N examples, and the softmax function transforms them to normalized
+   * probabilities.
+   *
+   * This can be interpreted as a generalization of the sigmoid
+   * function to multiple classes.
+   *
+   *   `probs_ij = e^scores_ij / sum(e^scores_i)`
+   *
+   * Inputs:
+   *  - scores: Inputs, of shape (N, C*HW).
+   *  - C: Number of channels or output labels.
+   *  - H: Image height
+   *  - W: Image width
+   *
+   * Outputs:
+   *  - probs: Outputs, of shape (N, C*HW).
+   */
+  # For numerical stability, we subtract the max score of an example from all scores for that
+  # example.  This is equivalent to the original formulation:
+  # e^scores_i / sum(e^scores_i) == C*e^scores_i / C*sum(e^scores_i)
+  #                              == e^(scores_i+log(C)) / sum(e^(scores_i+log(C))
+  # set log(C) = -max(scores_i):
+  #                              == e^(scores_i-max(scores_i)) / sum(e^(scores_i-max(scores_i))
+  N = nrow(scores)
+  scores = matrix(scores, rows=N*C, cols=H*W)
```

Comment:
What I was trying to say in the other PR is, we can convert from (N, C*H*W) to (C, N*H*W) with 'efficient' DML. What I mean by 'efficient' is without the use of loops. Copying the example from my earlier comment, this would convert
`20 21 22 23 24 25 26 27`
`30 31 32 33 34 35 36 37` 
to 
`20 21 22 23 30 31 32 33`
`24 25 26 27 34 35 36 37`
where N=C=H=W=2

What I understand from your comment is that you want (N, C*H*W) -> (N*H*W, C). In the case of the above example, it seems you want
`20 24`
`21 25`
`22 26`
`23 27`
`30 34`
`31 35`
`32 36`
`33 37`
which is the previous transformation followed by a `t()` in DML. If this is correct, then yes I can add the util.

On the other hand, its becoming increasingly clear to me that we need a transform4d. In which case, I would just add the builtin function and not bother with these utilities (no point doing redundant work).